examples/solids/qfunctions/hyperFS.h

```diff
@@ -0,0 +1,369 @@
+    // C = F^T * F (^T  means Transpose)
+    const CeedScalar C00 = F[0][0]*F[0][0] + F[0][1]*F[0][1] + F[0][2]*F[0][2];
+    const CeedScalar C01 = F[0][0]*F[1][0] + F[0][1]*F[1][1] + F[0][2]*F[1][2];
+    const CeedScalar C02 = F[0][0]*F[2][0] + F[0][1]*F[2][1] + F[0][2]*F[2][2];
+    const CeedScalar C11 = F[1][0]*F[1][0] + F[1][1]*F[1][1] + F[1][2]*F[1][2];
+    const CeedScalar C12 = F[1][0]*F[2][0] + F[1][1]*F[2][1] + F[1][2]*F[2][2];
+    const CeedScalar C22 = F[2][0]*F[2][0] + F[2][1]*F[2][1] + F[2][2]*F[2][2];
+    // *INDENT-OFF*
+    const CeedScalar C[3][3] = {{C00, C01, C02},
+                                {C01, C11, C12},
+                                {C02, C12, C22}
+                               };
+    // *INDENT-ON*
+
+    // Compute C^(-1) : C-Inverse
+    const CeedScalar A00 = C[1][1]*C[2][2] - C[1][2]*C[2][1];
+    const CeedScalar A01 = C[0][2]*C[2][1] - C[0][1]*C[2][2];
+    const CeedScalar A02 = C[0][1]*C[1][2] - C[0][2]*C[1][1];
+    const CeedScalar A11 = C[0][0]*C[2][2] - C[0][2]*C[2][0];
+    const CeedScalar A12 = C[0][2]*C[1][0] - C[0][0]*C[1][2];
+    const CeedScalar A22 = C[0][0]*C[1][1] - C[0][1]*C[1][0];
+    const CeedScalar detC = C[0][0]*A00 + C[1][0]*A01 + C[2][0]*A02;
+    const CeedScalar C_inv00 = A00/detC, C_inv01 = A01/detC, C_inv02 = A02/detC,
+                     C_inv11 = A11/detC, C_inv12 = A12/detC, C_inv22 = A22/detC;
+    // *INDENT-OFF*
+    const CeedScalar C_inv[3][3] = {{C_inv00, C_inv01, C_inv02},
+                                    {C_inv01, C_inv11, C_inv12},
+                                    {C_inv02, C_inv12, C_inv22}
+                                   };
+    // *INDENT-ON*
+
+    // deltaE - delta Green-Lagrange strain tensor
+    CeedScalar deltaE[3][3];
+    for (CeedInt j = 0; j < 3; j++)
+      for (CeedInt k = 0; k < 3; k++) {
+        deltaE[j][k] = 0;
+        for (CeedInt m = 0; m < 3; m++)
+          deltaE[j][k] += (graddeltau[m][j]*F[m][k] +
+                           F[m][j]*graddeltau[m][k])/2.;
+      }
+
+    // Compute the Second Piola-Kirchhoff (S)
+    const CeedScalar llnj_m = lambda*log(detF) - mu;
+    const CeedScalar S00 = mu + llnj_m*C_inv[0][0], S01 = llnj_m*C_inv[0][1],
+                     S02 = llnj_m*C_inv[0][2], S11 = mu + llnj_m*C_inv[1][1],
+                     S12 = llnj_m*C_inv[1][2], S22 = mu + llnj_m*C_inv[2][2];
+    // *INDENT-OFF*
+    CeedScalar S[3][3] = {{S00, S01, S02},
+                          {S01, S11, S12},
+                          {S02, S12, S22}
+                         };
+    // *INDENT-ON*
+
+    // deltaS = dSdE:deltaE
+    CeedScalar deltaS[3][3];
+    for (CeedInt j = 0; j < 3; j++)
+      for (CeedInt k = 0; k < 3; k++) {
+        deltaS[j][k] = 0;
+        for (CeedInt m = 0; m < 3; m++)
+          for (CeedInt n = 0; n < 3; n++)
+            deltaS[j][k] += lambda*C_inv[m][n]*deltaE[m][n]*C_inv[j][k] -
+                            2.*llnj_m*C_inv[j][m]*deltaE[m][n]*C_inv[n][k];
```

Comment:
`C^{-1} : dE` can be hoisted out of all these loops. I would also evaluate `C^{-1} dE C^{-1} = C^{-1} (dE C^{-1})` -- i.e., two matrix products instead of one. We know every matrix is symmetric so we could use Voigt notation for everything (but we can leave that optimization for later).

Reply:
Thanks